### DISCUSSION TITLE
Trying to get Extensions registered in Startup for SyncFusion BoldReports

### DISCUSSION THREAD

**Discussion by mixlplix** [oqtane/oqtane.framework#1624] [Q&A]
We are implementing SyncFusion BoldReports, and have been able to do so successfully by creating a class that implements and IServerStartup and IHostResources.  There are additional features we need to implement, namely adding new data sources.  The data source is a WebAPI datasource that is located in a different class library.  In an brand new project, we were able to implement this without an issue.  Here's the documentation.

https://help.boldreports.com/embedded-reporting/how-to/configure-webapi-data-extension-for-report-designer/

Basically, all you need is this line:

ReportConfig.DefaultSettings = new ReportSettings().RegisterExtensions(new List<string> { "BoldReports.Data.WebData" });

For some reason this does not work in Oqtane.  We can not get the web api data sources to register.  Attached is the class we have implemented.  Any suggestions on why this would not work?

[BoldReportService .txt](https://github.com/oqtane/oqtane.framework/files/7024470/BoldReportService.txt)

**Comment by sbwalker** [comment]
Generally these types of issues are related to assembly loading. In a standard .NET Core project all dependencies are statically linked so the system knows what needs to be loaded at startup. In a modular framework like Oqtane there is no static linking - everything is loaded dynamically. So my question would be how your datasource Web API class library is being loaded at runtime?

**Reply by mixlplix** [reply]
Thanks @sbwalker, I have not been able to get that answer out of the developers.  They just came back with a version that worked in Oqtane by adding the packages and config right into the Oqtane startup class, which is not an acceptable solution.  I will try again, but do you have a suggestion on how the class library should be loading at runtime?

**Reply by sbwalker** [reply]
@mixlplix there is a lot of "magic" that happens during startup in a .NET Core application. In a standard .NET Core project this magic is made possible by having all dependencies statically linked in the project file. Oqtane is modular so it cannot rely on this magic. If I explain how Oqtane handles startup it might give you some hints on how to solve your problem. During startup the system needs to know which assemblies in the /bin need to be loaded ( they are not loaded by default like they were in .NET Framework ). For Oqtane it relies on a naming convention where assemblies that have "oqtane" as part of their filename are loaded programmatically using AssemblyLoadContext. If the assemblies which are loaded contain dependencies to other assemblies, AssemblyLoadContext will attempt to load those as well ( which means they need to be located in the /bin ). In the case of assemblies which contain Controllers ( ie. Web APIs ), there is an additional step required to ensure the routes are added to the route table. This is done using ApplicationPartFactory. I do not know any details about your project but one possible solution may be to simply add "oqtane" to the filename of the assembly for your Web API class library ( which would allow Oqtane to recognize it and load it during startup ).

**Reply by mixlplix** [reply]
Here's the code from SyncFusion that loads the Assemblies in the "RegisterExtensions" function above

public ReportSettings RegisterExtensions(List<string> assemblies)
        {
            try
            {
                var localPath = System.IO.Path.GetDirectoryName(new System.Uri(System.Reflection.Assembly.GetExecutingAssembly().CodeBase).LocalPath);
 
                foreach (var item in assemblies)
                {
                    Assembly _loadAssembly = null;
                    var assembly = item.EndsWith(".dll", StringComparison.InvariantCultureIgnoreCase) ? item : string.Format(CultureInfo.InvariantCulture, "{0}.dll", item);
                    try
                    {
                        _loadAssembly = Assembly.LoadWithPartialName(assembly);
                    }
                    catch (Exception ex)
                    {
                        RDL.Internal.ReportUtil.LogUnknownException("RE0027", "", ex, "RegisterExtensions", "ReportSettings");
                    }
 
                    try
                    {
                        var assemblyPath = System.IO.Path.Combine(localPath, assembly);
                        if (_loadAssembly == null && System.IO.File.Exists(assembly))
                        {
                            _loadAssembly = Assembly.LoadFrom(assembly);
                        }
                        else if (_loadAssembly == null && System.IO.File.Exists(assemblyPath))
                        {
                            _loadAssembly = Assembly.LoadFrom(assemblyPath);
                        }
 
                        if (_loadAssembly != null)
                        {
                            var dataTypes = _loadAssembly.GetTypes().Where(t => t.IsClass);
                            if (dataTypes.Any())
                            {
                                foreach (Type data in dataTypes)
                                {
                                    var interfaces = data.GetInterfaces();
                                    var hasDataExtension = interfaces.Contains(typeof(BoldReports.Data.IDataExtension));
                                    var hasItemExtension = interfaces.Contains(typeof(BoldReports.RDL.Data.ICustomReportItem));
 
                                    if (hasDataExtension || hasItemExtension)
                                    {
                                        var customAttributes = data.GetCustomAttributes(false);
                                        if (customAttributes.Any())
                                        {
                                            var dataAttributes = customAttributes.Where(t => t.GetType() == typeof(ExtensionConfig));
                                            foreach (var dataAttr in dataAttributes)
                                            {
                                                Extension _extension = new Extension();
                                                _extension.AssemblyLocation = new System.Uri(_loadAssembly.CodeBase).LocalPath;
                                                _extension.ClassType = data.FullName;
                                                _extension.AssemblyName = assembly;
                                                _extension.AssemblyName = _loadAssembly.GetName().Name;
                                                _extension.IsBuiltInExtension = false;
                                                _extension.ExtensionType = hasDataExtension ? ExtensionType.Datasource : ExtensionType.ReportItem;
 
                                                var dataAttribute = dataAttr as ExtensionConfig;
                                                if (dataAttribute != null)
                                                {
                                                    _extension.Name = dataAttribute.Name;
                                                    if (dataAttribute.Visibility && !string.IsNullOrEmpty(_extension.Name))
                                                    {
                                                        if (this.Extensions == null)
                                                        {
                                                            this.Extensions = new List<Extension>();
                                                        }
 
                                                        this.Extensions.Add(_extension);
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                    catch(Exception ex)
                    {
                        string message = string.Format("Failed to Load Data extension {0} assembly, Exception: {1}", assembly, ex.Message);
                        BoldReports.RDL.Internal.ReportUtil.LogError("", message, ex, "", null, "RegisterExtensions", "ReportSettings");
                    }
                }
            }
            catch (Exception ex)
            {
                BoldReports.RDL.Internal.ReportUtil.LogError("", ex.Message, ex, "", null, "RegisterExtensions", "ReportSettings");
            }
 
            return this;
        }

**Reply by sbwalker** [reply]
This method accepts a List of assemblies as an input parameter - however I am not sure where it obtains this list from? It appears to be a list of strings so I would hope that the calling code is intelligent enough to only include the assemblies that are directly related to Syncfusion - and not the entire list of assemblies in the /bin. Either way, this code loads the assemblies and then inspects them for specific interfaces (ie. IDataExtension and ICustomReportItem) and builds a ReportSettings object which it returns. It does not load the assemblies into the app domain and it does not mention ApplicationPartFactory so it does not appear to take responsibility for registering the API routes. I am guessing it relies on .NET Core for this - however since your API class library is not statically linked to the the Oqtane solution, .NET Core will not be aware of the class library and therefore the API routes will not be registered. This is exactly why Oqtane must take responsibility for loading and registering assemblies as I described in my earlier message. The other problem you may run into is related to any dependencies that either BoldReports or your class library relies upon. You need to ensure that you copy ALL dependent DLLs to the Oqtane /bin as there would be no way for Oqtane to find them otherwise ( note that in an app with static linking such as via a Nuget package reference - .NET Core will determine all of the dependencies and download them to the /bin automatically behind the scenes ).
